pkg/webhook/server.go

```diff
@@ -0,0 +1,202 @@
+	// This maps to field .webhooks.clientConfig.service
+	// https://github.com/kubernetes/api/blob/183f3326a9353bd6d41430fc80f96259331d029c/admissionregistration/v1beta1/types.go#L260
+	Service *apitypes.NamespacedName
+	// Labels is the selector of the service
+	Labels map[string]string
+
+	// Provisioner provisions certificates for the admission webhook server.
+	// This is optional. If unspecified, it will be defaulted to use a self-signed certificate.
+	CertProvisioner *cert.Provisioner
+}
+
+// Server is an admission webhook server that can serve traffic and
+// generates related k8s reources for deploying.
+type Server struct {
+	// Name is the name of server
+	Name string
+
+	sMux *http.ServeMux
+	// registry maps a path to a http.Handler.
+	registry map[string]Webhook
+
+	// webhookConfigurations is a list of mutatingWebhookConfiguration and/or validatingWebhookConfiguration.
+	// This list is populated during server bootstrapping.
+	webhookConfigurations []runtime.Object
+
+	// ServerOptions contains options for configuring the admission server.
+	ServerOptions
+
+	once sync.Once
+}
+
+// Webhook defines the basics that a webhook should support.
+type Webhook interface {
+	// GetName returns the name of the webhook.
+	GetName() string
+	// GetPath returns the path that the webhook registered.
+	GetPath() string
+	// GetType returns the Type of the webhook.
+	// e.g. mutating or validating
+	GetType() types.WebhookType
+	// Handler returns a http.Handler for the webhook.
+	Handler() http.Handler
+	// SetKVMap sets the KVMap.
+	SetKVMap(map[string]interface{})
+}
+
+// NewServer creates a new admission webhook server.
+func NewServer(name string, mgr manager.Manager, ops ServerOptions) (*Server, error) {
+	as := &Server{
+		Name:          name,
+		ServerOptions: ops,
+		sMux:          http.NewServeMux(),
+		registry:      map[string]Webhook{},
+	}
+
+	return as, mgr.Add(as)
+}
+
+// DefaultServer returns an Server and defaults all possible fields.
+var DefaultServer = &Server{}
+
+// Register registers webhook(s) in the server
+func (s *Server) Register(webhooks ...Webhook) error {
+	for i, webhook := range webhooks {
+		_, found := s.registry[webhook.GetPath()]
+		if found {
+			return fmt.Errorf("can't register duplicate path: %v", webhook.GetPath())
+		}
+		s.registry[webhook.GetPath()] = webhooks[i]
+		s.sMux.Handle(webhook.GetPath(), webhook.Handler())
+		webhook.SetKVMap(s.KVMap)
+	}
+	return nil
+}
+
+var _ manager.Runnable = &Server{}
+
+// Start runs the server.
+func (s *Server) Start(stop <-chan struct{}) error {
+	_, err := s.bootstrap(false)
+	if err != nil {
+		return err
+	}
+
+	srv := &http.Server{
+		Addr:    fmt.Sprintf(":%v", s.Port),
+		Handler: s.sMux,
+	}
+	errCh := make(chan error)
+	serveFn := func() {
+		err := srv.ListenAndServeTLS(path.Join(s.CertDir, writer.ServerCertName), path.Join(s.CertDir, writer.ServerKeyName))
+		errCh <- err
+	}
+
+	for {
+		// TODO(mengqiy): add jitter to the timer
```

Comment:
Could use https://godoc.org/k8s.io/apimachinery/pkg/util/wait#Jitter for this.

Reply:
Added a TODO with your suggested lib.
Will address that in a followup PR.